### PR TITLE
refactor: casl factory - attachments

### PR DESCRIPTION
**Pull request by HayenNico**
## Description
Subsection of PR #2748 for attchments. PR depends on #2759 to be merged first.

This unifies the attachmentEndpointAccess and attachmentInstanceAccess functions in CaslAbilityFactory, removes instance-level Action elements and adjusts the affected controller to accommodate the change. The attachment-specific code is extracted into a separate module.

## Changes:

* Replace CaslAbilityFactory.attachmentInstanceAccess and CaslAbilityFactory.attachmentEndpointAccess with one function CaslAbilityFactory.attachmentAccess
* Code for CaslAbilityFactory.attachmentAccessis factored out into new module AttachmentAbility
* Remove all instance-level attachment Action elements, rename endpoint-level actions to cover both
* Adjust endpoint and instance auth logic in attachment controller

## Tests included

- [x] Included for each change/fix?
- [x] Passing? 

## Documentation
- [x] swagger documentation updated (required for API changes)
- [x] official documentation updated

## Summary by Sourcery

Unify attachment authorization into a single CASL ability and simplify attachment actions used across the API.

Enhancements:
- Replace separate attachment endpoint and instance accessors with a unified attachmentAccess ability in CaslAbilityFactory.
- Consolidate attachment endpoint and instance actions into a single set of attachment CRUD actions in the Action enum.
- Update the attachments v4 controller to rely on the unified attachmentAccess ability and new attachment actions for all route and instance-level permission checks.

### CODE REVIEW

File: src/attachments/attachments.v4.controller.ts [hunk 1]
```diff
@@ -215,7 +206,7 @@ export class AttachmentsV4Controller {
   // GET /attachments
   @UseGuards(PoliciesGuard)
   @CheckPolicies("attachments", (ability: AppAbility) =>
-    ability.can(Action.AttachmentReadEndpoint, Attachment),
+    ability.can(Action.AttachmentRead, Attachment),
```
[the remark this answers is not in the document]
Comment (HayenNico) [reply src/attachments/attachments.v4.controller.ts:217]: Unauthenticated user requests are properly handled in `AttachmentsV4Controller.addAccessBasedFilters` outside of casl. On the endpoint check, unauthenticated users *should* be let through to reproduce the current auth behavior.

**Review by nitrosx — APPROVED** [review]
Code looks good.
Please review the AI generated assessment and see what is applicable

**Comment by nitrosx** [comment]
# Code Review: refactor-casl-factory-attachments Branch

## Assessment

This branch introduces a significant refactoring of the CASL (Authorization) system for attachments. The changes consolidate and simplify the permission logic by:

1. **Creating a dedicated AttachmentAbility class** (`src/casl/abilities/attachments.ability.ts`) that encapsulates all attachment-related permission logic in a single, focused file.

2. **Simplifying action naming** in `action.enum.ts` by removing the redundant distinction between endpoint and instance actions for attachments:
   - Removed: `AttachmentCreateEndpoint`, `AttachmentReadEndpoint`, `AttachmentUpdateEndpoint`, `AttachmentDeleteEndpoint`
   - Removed: `AttachmentCreateInstance`, `AttachmentReadInstance`, `AttachmentUpdateInstance`, `AttachmentDeleteInstance`
   - Added: `AttachmentCreate`, `AttachmentRead`, `AttachmentUpdate`, `AttachmentDelete`

3. **Refactoring the factory** (`casl-ability.factory.ts`) to delegate attachment permissions to the new `AttachmentAbility` class, removing ~178 lines of duplicated and scattered logic.

4. **Updating the controller** (`attachments.v4.controller.ts`) to use the new unified action names and simplified permission checking methods.

5. **Registering the new ability** in the CASL module (`casl.module.ts`) as a provider.

The refactoring follows a pattern of extracting domain-specific abilities into separate classes, which improves maintainability and separation of concerns. The permission logic itself appears to be functionally equivalent, with some improvements in handling unauthenticated users.

## Improvement needed

### 1. Inconsistent Action Naming Pattern
The new attachment actions (`AttachmentCreate`, `AttachmentRead`, `AttachmentUpdate`, `AttachmentDelete`) follow a simpler naming pattern compared to other entities (e.g., datasets have both endpoint actions like `DatasetCreate` and instance actions like `DatasetCreateOwner`, `DatasetCreateAny`). This inconsistency could lead to confusion. Consider either:
- Keeping the endpoint/instance distinction for consistency with other entities
- Refactoring all entities to use the simplified pattern

### 2. Missing Type Safety in AttachmentAbility
The `AttachmentAbility.buildAbility()` method accepts `JWTUser | null` but the factory's `attachmentAccess()` method passes `JWTUser` directly. This type mismatch could cause issues when `user` is null in the controller context. The method should handle null users consistently.

### 3. Logic Gap in Filter Building
In `attachments.v4.controller.ts`, the `addAccessBasedFilters` method has a potential logic gap:
```typescript
if (!user) {
  // handles unauthenticated
} else if (!canViewAny && canView) {
  // handles authenticated with limited permissions
}
```
There's no `else` case to handle authenticated users who cannot view any attachments and cannot view specific attachments. This could result in no filters being applied when they should be.

### 4. Incomplete Migration
The refactoring only addresses attachments. Other entities in the codebase (datablocks, datasets, etc.) still use the old pattern with separate endpoint and instance access methods. This partial migration creates inconsistency in the codebase architecture.

### 5. Testing Coverage
There's no evidence of updated tests to verify that the refactored permission logic behaves identically to the previous implementation. The changes affect security-critical functionality and should have comprehensive test coverage.

### 6. Code Organization
The new `AttachmentAbility` class is placed in a new `abilities` subdirectory, but it's the only ability class there. Consider whether this pattern should be applied to other entities as well, or if the class should be placed at the same level as the factory.

## Verdict

**The refactoring is necessary and beneficial** for improving code maintainability and reducing duplication. The extraction of attachment-specific permission logic into a dedicated class is a good architectural decision that follows separation of concerns principles.

**However, the implementation has issues that need to be addressed** before merging:
1. The inconsistent action naming pattern should be resolved for codebase cohesion
2. The type safety issue with null users should be fixed
3. The logic gap in filter building needs to be addressed
4. Comprehensive tests should be added to verify the refactored permissions work correctly
5. Consider whether to complete the migration for other entities or revert to maintain consistency

**Recommendation**: Address the identified issues, add comprehensive tests, and consider whether to extend this pattern to other entities for consistency. The architectural approach is sound, but the execution needs refinement to maintain the existing security guarantees and code quality standards.

**Comment by nitrosx** [comment]
@HayenNico Improvements #1, #4, and #6 are not relevant.

**Comment by nitrosx** [comment]
# Code Review: refactor-casl-factory-attachments Branch

## Summary of Changed Code

The refactoring extracts attachment-specific CASL permission logic from the monolithic `CaslAbilityFactory` into a dedicated `AttachmentAbility` class, while simplifying the action naming convention.

**Files Modified:**
- `src/casl/abilities/attachments.ability.ts` (new) - Encapsulates all attachment permission logic
- `src/casl/action.enum.ts` - Replaces 8 attachment actions with 4 simplified ones
- `src/casl/casl-ability.factory.ts` - Delegates to `AttachmentAbility`, removes 178 lines
- `src/casl/casl.module.ts` - Registers new `AttachmentAbility` provider
- `src/attachments/attachments.v4.controller.ts` - Uses new action names and simplified factory method

## Logic Correctness

**✅ CORRECT**: The permission logic is functionally equivalent to the original implementation.

- **Unauthenticated users**: Can only read published attachments (same as before)
- **Authenticated users**: Always get read access for attachments they own, have access to, or are published
- **ATTACHMENT_GROUPS**: Can create/update/delete only own attachments
- **ATTACHMENT_PRIVILEGED_GROUPS**: Can create any, update/delete only own
- **ADMIN_GROUPS**: Full access including `AccessAny`
- **DELETE_GROUPS**: Can delete any attachment

The new implementation uses additive `if` statements instead of `else if`, which is more correct as users can belong to multiple groups.

## Edge Cases

**✅ HANDLED**: All edge cases are properly addressed:

1. **Null user**: `AttachmentAbility.buildAbility()` accepts `JWTUser | null` and handles unauthenticated access
2. **No permissions**: Default read access for published attachments applies
3. **Multiple group memberships**: Additive permission logic correctly handles users in multiple groups
4. **Admin users**: Full access bypasses all restrictions as expected

## Logic Gaps

**⚠️ MINOR ISSUE**: In `addAccessBasedFilters()`, the condition `!canViewAny && canView` could be simplified.

- For authenticated users, `canView` is always true (all authenticated users get read permissions)
- The condition could be just `!canViewAny` for clarity
- Current logic is functionally correct but potentially confusing

## Unreachable Code

**✅ NONE**: No unreachable code identified.

- All permission branches are reachable for appropriate user types
- All filter conditions are properly guarded
- No redundant permission rules

## Does It Make Sense?

**✅ YES**: The refactoring is well-designed and beneficial:

1. **Separation of Concerns**: Attachment permissions are now isolated in their own class
2. **Code Reuse**: Single permission definition used for both endpoint and instance checks
3. **Maintainability**: 178 fewer lines, clearer structure, easier to modify
4. **Consistency**: Follows the pattern of extracting domain-specific abilities

## Verdict

**APPROVE WITH MINOR NOTES**: The changes are logically correct, maintain the same security model, handle all edge cases properly, and represent a significant improvement in code organization. The only recommendation is to simplify the filter condition for clarity, but the current implementation is functionally sound.

**Comment by nitrosx** [comment]
# Security Review: refactor-casl-factory-attachments Branch

## 1. Injection Vulnerabilities: ❌ NONE

**Analysis:**
- All MongoDB conditions use CASL's type-safe ability builder with values from validated JWT tokens
- User data (`currentGroups`, `email`) is sourced from already-authenticated JWT, not raw user input
- No dynamic query construction from user-supplied strings
- No use of `eval`, `$where`, or other MongoDB injection vectors
- Values used in conditions: `user.currentGroups`, `user.accessGroups`, `user.email` - all from authenticated JWT

**Example of safe usage:**
```typescript
// attachments.ability.ts
const ifOwner = { ownerGroup: { $in: user.currentGroups } };
can(Action.AttachmentRead, Attachment, ifOwner);
```
→ `currentGroups` comes from validated JWT, not user input.

## 2. Sensitive User Data Exposure: ❌ NONE

**Analysis:**
- **No new endpoints** introduced
- **No data serialization changes** — only authorization logic refactored
- **Same data returned** to clients; only permission checks changed
- Public/private data visibility controlled by `isPublished`, `ownerGroup`, `accessGroups` — unchanged
- No new user fields exposed in responses

**Verdict:** The refactoring does not expose any additional user data.

## 3. Insecure API Usage: ❌ NONE

**Analysis:**
- **No new external API calls** introduced
- **No hardcoded credentials, tokens, or secrets**
- **No changes to configuration** for external services
- All changes are internal to the authorization layer
- No new HTTP clients or external service integrations

**Verdict:** No insecure API usage patterns introduced.

## 4. Authentication Bypass: ❌ NONE

**Deep Analysis:**

| Attack Vector | Old Code | New Code | Risk |
|--------------|----------|----------|------|
| **Unauthenticated read** | Allowed public attachments via `AttachmentReadInstance` with `isPublished: true` | Allowed via `can(AttachmentRead, Attachment, { isPublished: true })` | ✅ Same |
| **Admin escalation** | `AccessAny` granted to admin | `AccessAny` granted to admin | ✅ Same |
| **Group-based access** | Separate actions with OR conditions | Single action with OR conditions | ✅ Equivalent |
| **Permission accumulation** | Multiple `can()` calls with OR | Multiple `can()` calls with OR | ✅ Same |

**CASL Semantics:**
- Multiple `can(Action.AttachmentRead, ...)` calls create **OR conditions**
- `cannot()` rules have higher priority than `can()` (preserved in new code)
- The new implementation uses additive `if` statements instead of `else if`, which is more correct for users in multiple groups

**Verdict:** No authentication bypass possible. The refactoring maintains the same security model.

## Final Verdict: ALL SECURE ✅

| Security Concern | Status | Notes |
|-----------------|--------|-------|
| Injection Vulnerabilities | ❌ NONE | Type-safe queries, validated JWT sources |
| Sensitive Data Exposure | ❌ NONE | No new endpoints or serialization changes |
| Insecure API Usage | ❌ NONE | No external API changes |
| Authentication Bypass | ❌ NONE | Equivalent permission logic; no gaps |

**Conclusion:** The refactoring maintains the same security posture while improving code organization. No security vulnerabilities introduced.

**Comment by nitrosx** [comment]
# PR Review Report - refactor-casl-factory-attachments

Generated: 2026-08-11 11:05:16
Branch: refactor-casl-factory-attachments
Base: master

---

## Overview

### 1. Logic Correctness Verification
The refactoring maintains logical equivalence with the original implementation while improving code organization. The authorization logic has been consolidated from two separate methods (`attachmentEndpointAccess` and `attachmentInstanceAccess`) into a single `AttachmentAbility.buildAbility()` method that handles both endpoint and instance-level permissions using the same permission rules.

**Commits responsible**: `fdb3e3df` (extract attachments casl ability), `60c116d8` (unauthenticated user handling)

### 2. Edge Cases Handling
All edge cases are properly handled:
- **Unauthenticated users**: Can only read public attachments (isPublished: true) - correctly implemented in `AttachmentAbility.buildAbility()` line 39
- **Null/undefined accessGroups**: Safely handled with optional chaining (`this.accessGroups?.attachment?.includes(g)`) and nullish coalescing (`?? {}`) - fixed in commit `1f174ccc`
- **Null user**: Explicitly checked and returns early with read-only public access - implemented in lines 41-46
- **Empty user.currentGroups**: Safe iteration with `.some()` which returns false for empty arrays

### 3. Code Touched Summary
The changes touch the CASL authorization layer for attachments, refactoring it from a monolithic factory-based approach to a modular ability-based architecture:

- **Authorization Logic**: Extracted from `CaslAbilityFactory` into dedicated `AttachmentAbility` class
- **Action Definitions**: Simplified from 8 endpoint/instance-specific actions to 4 semantic actions
- **Controller Integration**: Updated to use unified action names and access methods
- **Documentation**: Added comprehensive authorization model documentation

### 4. Changes Assessment
The changes make architectural sense by:
- Following the Single Responsibility Principle (each ability class handles one resource type)
- Reducing code duplication (eliminates ~180 lines of duplicated permission logic)
- Improving maintainability (changes to attachment permissions only need to be made in one place)
- Enhancing consistency (same pattern as datasets refactoring in related branch)

**Verdict**: Changes are well-structured and represent a legitimate architectural improvement.

### 5. Unreachable Code
No unreachable code identified. The refactoring removed the original `attachmentEndpointAccess` and `attachmentInstanceAccess` methods entirely, replacing them with the unified approach. All remaining code paths are accessible and logical.

---

## Code Changes

### Summary of Changes

| File | Change Type | Lines Changed | Purpose |
|------|-------------|---------------|---------|
| `src/casl/abilities/attachments.ability.ts` | New | +111 | New dedicated ability class for attachments |
| `docs/developer-guide/authorization/attachments.md` | New | +70 | Authorization documentation |
| `src/casl/action.enum.ts` | Modified | -18 | Simplified action definitions |
| `src/casl/casl-ability.factory.ts` | Modified | -178 | Removed duplicate methods, added AttachmentAbility integration |
| `src/casl/casl.module.ts` | Modified | +2 | Added AttachmentAbility provider |
| `src/attachments/attachments.v4.controller.ts` | Modified | ~50 | Updated action references and method calls |
| `src/casl/casl-ability.factory.spec.ts` | Modified | +4 | Added AttachmentAbility to test setup |

### Detailed Analysis

#### 1. New AttachmentAbility Class (`attachments.ability.ts`)
**Assessment**: Necessary and well-implemented
- Encapsulates all attachment authorization logic in one place
- Follows the same pattern as `DatasetAbility` from the related refactoring
- Properly handles all user types (unauthenticated, authenticated, privileged groups)
- Uses CASL's MongoAbility for MongoDB-style condition matching

**Improvement Needed**:
- Missing unit tests for the `AttachmentAbility` class itself
- Could benefit from more detailed JSDoc comments for complex permission logic

**Verdict**: Excellent architectural decision with proper implementation.

#### 2. Action Enum Simplification (`action.enum.ts`)
**Assessment**: Necessary and beneficial
- Removed 8 endpoint/instance-specific actions (`AttachmentCreateEndpoint`, `AttachmentReadEndpoint`, etc.)
- Added 4 semantic actions (`AttachmentCreate`, `AttachmentRead`, `AttachmentUpdate`, `AttachmentDelete`)
- Reduces cognitive load and improves consistency with other resource types

**Improvement Needed**: None identified

**Verdict**: Clean simplification that improves code clarity.

#### 3. CaslAbilityFactory Refactoring (`casl-ability.factory.ts`)
**Assessment**: Necessary for architectural consistency
- Removed `attachmentEndpointAccess` method (-67 lines)
- Removed `attachmentInstanceAccess` method (-111 lines)
- Added `attachmentAccess` method that delegates to `AttachmentAbility` (+4 lines)
- Updated endpoint accessors to use new method name
- Added `AttachmentAbility` as constructor dependency

**Improvement Needed**:
- The factory still contains significant code for other resources; consider extracting more abilities in future

**Verdict**: Successful extraction of concerns, significantly reduces factory complexity.

#### 4. Attachments Controller Updates (`attachments.v4.controller.ts`)
**Assessment**: Necessary to integrate refactored authorization
- Updated all action references from `*Endpoint` and `*Instance` to simple action names
- Changed `attachmentInstanceAccess` calls to `attachmentAccess`
- Updated `addAccessBasedFilters` to use new action names and handle unauthenticated users
- Removed redundant null checks (using nullish coalescing operator)

**Improvement Needed**:
- The `addAccessBasedFilters` method could be simplified further; the condition `!canViewAny && canView` on line 152 is somewhat redundant since `canViewAny` implies `canView`

**Verdict**: Successful integration with minor opportunity for further simplification.

### Overall Verdict
**EXCELLENT**: The refactoring achieves its goals of reducing duplication, improving maintainability, and following established architectural patterns. The changes are logically sound, well-integrated, and represent a significant improvement over the previous implementation.

---

## Security Review

### 1. Injection Vulnerabilities
**Finding**: None identified
- All database queries use parameterized MongoDB query objects (e.g., `{ ownerGroup: { $in: user.currentGroups } }`)
- No string concatenation or template literals used in query construction
- The `AttachmentAbility` class uses CASL's type-safe ability definitions

**Commits**: All changes maintain safe query construction patterns

### 2. Sensitive User Data Exposure
**Finding**: None identified
- Authorization checks are performed before any data access
- The refactoring maintains the same data access patterns as the original code
- No new endpoints or data exposure paths introduced
- User data (`currentGroups`, `email`) is only used for authorization decisions, not exposed

**Commits**: All authorization logic preserves data protection

### 3. Insecure API Usage
**Finding**: None identified
- CASL library is used correctly with proper ability definitions
- MongoAbility is configured with correct type parameters
- All permissions are checked using `ability.can()` before allowing operations
- No direct database access bypasses authorization

**Commits**: Proper CASL integration maintained

### 4. Authentication Bypass
**Finding**: None identified
- Unauthenticated users are explicitly handled (lines 41-46 in `attachments.ability.ts`)
- All permission checks properly validate user context
- The refactoring consolidates authorization logic, making it harder to accidentally bypass
- Endpoint guards (`@UseGuards(PoliciesGuard)`) remain in place
- Instance-level checks are performed in `checkPermissionsForAttachment` method

**Commits**: `60c116d8` specifically added proper unauthenticated user handling

**Security Verdict**: **SECURE** - No security vulnerabilities introduced. The refactoring actually improves security posture by consolidating authorization logic and reducing the attack surface through code duplication elimination.

---

## Test Coverage

### Current Coverage Assessment

| Component | Test Status | Coverage Level |
|-----------|-------------|----------------|
| `AttachmentAbility` | **MISSING** | No direct tests |
| `CaslAbilityFactory` | Partial | Tests updated for new constructor |
| `AttachmentsV4Controller` | Existing | Controller tests exist but may need updates |
| Integration | Existing | End-to-end authorization flows covered |

### Findings

1. **Missing Unit Tests**: No dedicated unit tests for `AttachmentAbility` class
   - The new class contains complex permission logic that should be unit tested
   - Different user types and group configurations should be tested

2. **Factory Tests Updated**: `casl-ability.factory.spec.ts` properly updated to include `AttachmentAbility` in constructor
   - However, tests only verify factory instantiation, not attachment-specific functionality

3. **Controller Tests**: Existing controller tests likely need review to ensure they work with new action names

### Suggestions for Improvement

#### High Priority
1. **Create `attachments.ability.spec.ts`** with tests for:
   - Unauthenticated user permissions (read public only)
   - Authenticated user with no special groups (read own/access/public)
   - User in ATTACHMENT_GROUPS (create/update/delete own)
   - User in ATTACHMENT_PRIVILEGED_GROUPS (create any, update/delete own)
   - User in ADMIN_GROUPS (full access)
   - User in DELETE_GROUPS (delete any)
   - Edge cases: null accessGroups, empty groups, undefined user

#### Medium Priority
2. **Update controller tests** to verify:
   - New action names work correctly
   - Permission checks integrate properly with controller methods
   - Filter application in `addAccessBasedFilters` works as expected

#### Test File Structure Suggestion
\`\`\`typescript
// src/casl/abilities/attachments.ability.spec.ts
describe('AttachmentAbility', () => {
  describe('Unauthenticated user', () => {
    it('can read public attachments');
    it('cannot create attachments');
    it('cannot update attachments');
    it('cannot delete attachments');
  });

  describe('Authenticated user', () => {
    describe('with no special groups', () => {
      it('can read own attachments');
      it('can read access group attachments');
      it('can read public attachments');
      it('cannot create attachments');
    });

    describe('in ATTACHMENT_GROUPS', () => {
      it('can create attachments with matching ownerGroup');
      it('can update own attachments');
      it('can delete own attachments');
    });

    // Similar sections for ATTACHMENT_PRIVILEGED_GROUPS, ADMIN_GROUPS, DELETE_GROUPS
  });
});
\`\`\`

---

## Security Examples

*This section is empty as no security vulnerabilities were identified in the Security Review.*

---

## Testing for Security Use Cases

*This section is empty as no security vulnerabilities were identified. However, the following testing approach is recommended for future security validation:*

### Recommended Security Testing Approach

1. **Authentication Bypass Testing**
   - Test unauthenticated access to all attachment endpoints
   - Verify only GET requests with public attachments succeed
   - Test with malformed or missing JWT tokens
   - File: `attachments.v4.controller.ts` lines 214-545

2. **Authorization Elevation Testing**
   - Test users with various group memberships
   - Verify users cannot perform actions outside their permission level
   - Test edge cases like empty groups, null groups
   - File: `attachments.ability.ts` lines 58-104

3. **Query Injection Testing**
   - Test with specially crafted group names containing MongoDB operators
   - Verify all queries use parameterized conditions
   - File: `attachments.ability.ts` lines 48-49 (condition definitions)

---

## Summary

### Key Findings

| Aspect | Status | Details |
|--------|--------|---------|
| **Architecture** | Excellent | Modular, maintainable, follows SRP |
| **Logic Correctness** | Verified | All permission logic maintains equivalence |
| **Edge Cases** | Handled | Null checks, empty arrays, undefined config |
| **Security** | Secure | No vulnerabilities introduced |
| **Test Coverage** | Inadequate | Missing AttachmentAbility unit tests |
| **Code Quality** | Improved | Reduced duplication, better organization |

### Consolidated Assessment

The `refactor-casl-factory-attachments` branch successfully refactors the attachment authorization system from a monolithic factory approach to a modular ability-based architecture. The changes:

1. **Are necessary**: The previous implementation had ~180 lines of duplicated permission logic split across two methods
2. **Are well-implemented**: Follow established patterns from the datasets refactoring, maintain logical equivalence
3. **Improve maintainability**: Single source of truth for attachment permissions
4. **Maintain security**: No security regressions, proper handling of all user types
5. **Need test coverage**: Unit tests for `AttachmentAbility` should be added before merging

### Recommendations

1. **Before Merge**: Add comprehensive unit tests for `AttachmentAbility` class
2. **Consider**: Extracting more resource abilities (datablocks, instruments, etc.) using the same pattern
3. **Document**: Update any architectural documentation to reflect the new pattern

### Final Verdict

**APPROVED WITH CONDITIONS**: The refactoring is architecturally sound, logically correct, and secure. The only blocking item is the addition of unit tests for the new `AttachmentAbility` class. Once adequate test coverage is added, this PR is ready for merge.

---

*Report generated by Mistral Vibe at 2026-08-11 11:05:16*

**Review by nitrosx — APPROVED** [review]
